core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -257,15 +257,6 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node, RelDataType
       return false;
     }
 
-    // No need to cast if the source type precedence list
```

Comment:
I found that you only changed the conversion rules between tinyint, int and bigint. Do you want to support more types for IN Sub-query?  For example:
```
  @Test void testInSubQuery() {
    sql("select * from t1 where t1_bigint in (select t1_decimal from t1)")
        .withExpand(false)
        .ok();
  }
```
> Learn from you.

Reply:
This PR focuses on integer types. We can add it in another PR.